<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Получение-и-обзор-данных" data-toc-modified-id="Получение-и-обзор-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Получение и обзор данных</a></span><ul class="toc-item"><li><span><a href="#Проверим-корректность-названия-столбцов." data-toc-modified-id="Проверим-корректность-названия-столбцов.-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Проверим корректность названия столбцов.</a></span></li><li><span><a href="#Общая-информация-о-данных-таблицы." data-toc-modified-id="Общая-информация-о-данных-таблицы.-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Общая информация о данных таблицы.</a></span></li><li><span><a href="#Проверим-величины-хранящиеся-в-таблице." data-toc-modified-id="Проверим-величины-хранящиеся-в-таблице.-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Проверим величины хранящиеся в таблице.</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Обработка-пропусков" data-toc-modified-id="Обработка-пропусков-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Обработка пропусков</a></span><ul class="toc-item"><li><span><a href="#Обработка-пропусков-в-столбце-total_income" data-toc-modified-id="Обработка-пропусков-в-столбце-total_income-2.1.1"><span class="toc-item-num">2.1.1&nbsp;&nbsp;</span>Обработка пропусков в столбце <code>total_income</code></a></span></li><li><span><a href="#Обработка-пропусков-в-столбце-days_employed" data-toc-modified-id="Обработка-пропусков-в-столбце-days_employed-2.1.2"><span class="toc-item-num">2.1.2&nbsp;&nbsp;</span>Обработка пропусков в столбце <code>days_employed</code></a></span></li></ul></li><li><span><a href="#Замена-типа-данных" data-toc-modified-id="Замена-типа-данных-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Замена типа данных</a></span></li><li><span><a href="#Исправление-ошибок-в-данных" data-toc-modified-id="Исправление-ошибок-в-данных-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Исправление ошибок в данных</a></span></li><li><span><a href="#Обработка-дубликатов" data-toc-modified-id="Обработка-дубликатов-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Обработка дубликатов</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Лемматизация" data-toc-modified-id="Лемматизация-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Лемматизация</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Категоризация-данных" data-toc-modified-id="Категоризация-данных-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Категоризация данных</a></span><ul class="toc-item"><li><span><a href="#Катетеризация-по-количеству-детей." data-toc-modified-id="Катетеризация-по-количеству-детей.-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Катетеризация по количеству детей.</a></span></li><li><span><a href="#Категоризация-по-доходам" data-toc-modified-id="Категоризация-по-доходам-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Категоризация по доходам</a></span></li><li><span><a href="#Категоризация-по-возрасту" data-toc-modified-id="Категоризация-по-возрасту-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Категоризация по возрасту</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Проверка-гипотез" data-toc-modified-id="Проверка-гипотез-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Проверка гипотез</a></span><ul class="toc-item"><li><span><a href="#Есть-ли-зависимость-между-наличием-детей-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-наличием-детей-и-возвратом-кредита-в-срок?-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Есть ли зависимость между наличием детей и возвратом кредита в срок?</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Вывод</a></span></li><li><span><a href="#Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?-5.3"><span class="toc-item-num">5.3&nbsp;&nbsp;</span>Есть ли зависимость между семейным положением и возвратом кредита в срок?</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-5.4"><span class="toc-item-num">5.4&nbsp;&nbsp;</span>Вывод</a></span></li><li><span><a href="#Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?-5.5"><span class="toc-item-num">5.5&nbsp;&nbsp;</span>Есть ли зависимость между уровнем дохода и возвратом кредита в срок?</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-5.6"><span class="toc-item-num">5.6&nbsp;&nbsp;</span>Вывод</a></span></li><li><span><a href="#Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?" data-toc-modified-id="Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?-5.7"><span class="toc-item-num">5.7&nbsp;&nbsp;</span>Как разные цели кредита влияют на его возврат в срок?</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-5.8"><span class="toc-item-num">5.8&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Общий вывод</a></span></li></ul></div>

# Исследование надёжности заёмщиков

**Заказчик** — кредитный отдел банка. 

**Цель исследования** — разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. 

**Входные данные от банка** — статистика о платёжеспособности клиентов.

**Результаты исследования** будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

**Ход исследования**

Исследование пройдёт в три этапа:
 1. Получение и обзор данных.
 2. Предобработка данных.
 3. Проверка гипотезы.


## Получение и обзор данных


In [91]:
#импортируем библиотеки необходимые для анализа данных

import pandas as pd
import numpy as np
import seaborn as sns

Прочитаем файл data_credit.csv и сохраним его в переменной clients.

In [2]:
clients = pd.read_csv('data_credit.csv')

Получение первых 10 строк таблицы.

In [3]:
clients.head(10)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

Можно заметить, что в колонке `education` присутствуют неоднородные данные. Следует привести к одному регистру.

In [4]:
clients.education.unique()

array(['высшее', 'среднее', 'Среднее', 'СРЕДНЕЕ', 'ВЫСШЕЕ',
       'неоконченное высшее', 'начальное', 'Высшее',
       'НЕОКОНЧЕННОЕ ВЫСШЕЕ', 'Неоконченное высшее', 'НАЧАЛЬНОЕ',
       'Начальное', 'Ученая степень', 'УЧЕНАЯ СТЕПЕНЬ', 'ученая степень'],
      dtype=object)

### Проверим корректность названия столбцов.


In [5]:
clients.columns

Index(['children', 'days_employed', 'dob_years', 'education', 'education_id',
       'family_status', 'family_status_id', 'gender', 'income_type', 'debt',
       'total_income', 'purpose'],
      dtype='object')

Названия корректны.

### Общая информация о данных таблицы.

In [6]:
clients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Рассмотрим полученную информацию подробнее.

Всего в таблице данных о 21525 клиентах с 12 признаками, 4 из которых относятся к количественным признакам, 8 – к категориальным.

Согласно документации к данным:

* `children` — количество детей в семье;
* `days_employed` — общий трудовой стаж в днях;  
* `dob_years` — возраст клиента в годах;
* `education` — уровень образования клиента;
* `education_id` — идентификатор уровня образования;
* `family_status` — семейное положение;
* `family_status_id` — идентификатор семейного положения;
* `gender` — пол клиента;
* `income_type` — тип занятости;
* `debt` — имел ли задолженность по возврату кредитов;
* `total_income` — ежемесячный доход;
* `purpose` — цель получения кредита.

Как можно заметить в колонках `days_employed` и `total_income` имеются пропущенные значения.

### Проверим величины хранящиеся в таблице.

In [7]:
clients.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06

Можно заметить, что в столбце `children`, `days_employed`, `dob_years` содержатся ненормальные для них значения, которые следует проверить и исправить:
    
1. Отрицательное значение в столбце `children`;
2. Максимальное количество детей в столбце `children` - 20;
3. Минимальное значение возраста клиента в `dob_years` - 0;
4. Отрицательные значение трудового стажа в `days_employed`;
5. Максимальное значение трудового стажа в `days_employed` составляет 401755.400475, что равняется ~ 1100 годам.

### Вывод

1. Пустые значения в `days_employed` и `total_income` свидетельствуют, что для некоторых данных доступна не вся информация. Причиной может быть отсутствие информации о трудовом стаже и о заработке.


2. Что касается ненормальности значений, то здесь возможны технические ошибки или ошибки при сборе данных.

**Столбец `children`:**

- Отрицательное значение в столбце  - скорее всего имеется в виду 1 ребенок, проверить частотность подобных значений, если ошибка нечастая, то взять по модулю или привести к 0.
- Максимальное количество детей в столбце children - необходимо проверить количество таких значений. Предполагаю, что имеется в виду 2 ребенка. 
    
**Столбец `dob_years`:**

- Минимальное значение возраста клиента в `dob_years` - выяснить причину ошибки и заменить на близкий к реальному возраст.
    
**Столбец `days_employed`.**

Уточнить у заказчика корректность собранных данных. Если такое не представляется возможным:
- Отрицательные значение трудового стажа в `days_employed` - можно предолжить, что данные введены некорректно, и необходимо привести данные к положительному значению взятием значения по модулю.
- Максимальное значение трудового стажа в days_employed привести к возможному значению.

3. Тип данных в `total_income` следует заменить на целочисленный.


4. Столбец `education` имеет неоднотипные значения - необходимо привести к общему виду.

## Предобработка данных


### Обработка пропусков



In [8]:
clients[clients['total_income'].isnull()].head()

children  days_employed  dob_years education  education_id  \
12         0            NaN         65   среднее             1   
26         0            NaN         41   среднее             1   
29         0            NaN         63   среднее             1   
41         0            NaN         50   среднее             1   
55         0            NaN         54   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   

    total_income                           purpose  
12           NaN                   сыграть свадьбу  
26           NaN                       образование  
29           NaN  строительство жилой недвижимости  
41           NaN  сделка с подержанным автомобилем  
55           NaN                   сыграть свадьбу

Подсчитаем количество пропусков

In [9]:
clients.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

Проверим взаимосвязаны ли пропущенные значения в `days_employed` и `total_income`

In [10]:
clients[(clients['days_employed'].isnull() == True) & (clients['total_income'].isnull() == True)].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2174 entries, 12 to 21510
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          2174 non-null   int64  
 1   days_employed     0 non-null      float64
 2   dob_years         2174 non-null   int64  
 3   education         2174 non-null   object 
 4   education_id      2174 non-null   int64  
 5   family_status     2174 non-null   object 
 6   family_status_id  2174 non-null   int64  
 7   gender            2174 non-null   object 
 8   income_type       2174 non-null   object 
 9   debt              2174 non-null   int64  
 10  total_income      0 non-null      float64
 11  purpose           2174 non-null   object 
dtypes: float64(2), int64(5), object(5)
memory usage: 220.8+ KB


Во всех 2174 ячейках, где пропущены `days_employed`, также пропущены и `total_income`. 

Найдем какую долю составляют пропущенные значения в каждом из столбцов с пропусками

In [11]:
missing_values = pd.DataFrame({'percent_of_nulls': clients.isnull().sum() / len(clients)})
missing_values['percent_of_nulls'] = missing_values['percent_of_nulls'].map(lambda x: '{:.1%}'.format(x))
missing_values.sort_values(by='percent_of_nulls', ascending = False).head()

percent_of_nulls
days_employed            10.1%
total_income             10.1%
children                  0.0%
dob_years                 0.0%
education                 0.0%

Доля пропущенных значений составляет 10% от всего колличества данных, что весьма существенно, поэтому логично будет заполнить пропуски медианным значением по каждому типу занятости и трудовому стажу.

#### Обработка пропусков в столбце `total_income`


Найдем медианные уровни дохода для каждого типа занятости и запишем это в переменную median_income

In [12]:
median_income = clients.groupby('income_type').agg({'total_income' : 'median'}).round(2)
median_income.reset_index()

income_type  total_income
0      безработный     131339.75
1        в декрете      53829.13
2      госслужащий     150447.94
3        компаньон     172357.95
4        пенсионер     118514.49
5  предприниматель     499163.14
6        сотрудник     142594.40
7          студент      98201.63

Заменим пропущенные значения ежемесячного дохода на медианы по типам занятости.

In [13]:
def fillna_total_income(dataframe):
    type_income = dataframe['income_type']
    if pd.isnull(dataframe['total_income']):
        return median_income.loc[type_income]
    else:
        return dataframe['total_income']

In [14]:
clients['total_income'] = clients.apply(fillna_total_income, axis = 1)

Пропуски в `total_income` заполнены. Пропусков в данных не осталось. Проверим это


In [15]:
clients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  object 
 11  purpose           21525 non-null  object 
dtypes: float64(1), int64(5), object(6)
memory usage: 2.0+ MB


#### Обработка пропусков в столбце `days_employed`


Для исправления пропущенных значений воспользуемся медианными значениями по возрасту.

In [16]:
median_days_employed = clients.groupby('dob_years').agg({'days_employed' : 'median'}).round(2)
median_days_employed.head()

days_employed
dob_years               
0               -1146.69
19               -724.49
20               -674.84
21               -618.73
22               -698.10

In [17]:
def fillna_days_employed(dataframe):
    age = dataframe['dob_years']
    if pd.isnull(dataframe['days_employed']):
        return median_days_employed.loc[age]
    else:
        return dataframe['days_employed']

In [18]:
clients['days_employed'] = clients.apply(fillna_days_employed, axis = 1)

Пропуски в `days_employed` заполнены. 
Проверим


In [19]:
clients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21525 non-null  int64 
 1   days_employed     21525 non-null  object
 2   dob_years         21525 non-null  int64 
 3   education         21525 non-null  object
 4   education_id      21525 non-null  int64 
 5   family_status     21525 non-null  object
 6   family_status_id  21525 non-null  int64 
 7   gender            21525 non-null  object
 8   income_type       21525 non-null  object
 9   debt              21525 non-null  int64 
 10  total_income      21525 non-null  object
 11  purpose           21525 non-null  object
dtypes: int64(5), object(7)
memory usage: 2.0+ MB


### Замена типа данных

В целом данные в таблице одержаться в подходящем типе данных. 
Переведем только данные в `total_income` и `days_employed` из 'float64' в целочисленное значение 'int'.

In [20]:
clients['days_employed'] = clients['days_employed'].astype('int64')
clients['total_income'] = clients['total_income'].astype('int64')

In [21]:
clients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21525 non-null  int64 
 1   days_employed     21525 non-null  int64 
 2   dob_years         21525 non-null  int64 
 3   education         21525 non-null  object
 4   education_id      21525 non-null  int64 
 5   family_status     21525 non-null  object
 6   family_status_id  21525 non-null  int64 
 7   gender            21525 non-null  object
 8   income_type       21525 non-null  object
 9   debt              21525 non-null  int64 
 10  total_income      21525 non-null  int64 
 11  purpose           21525 non-null  object
dtypes: int64(7), object(5)
memory usage: 2.0+ MB


Тип данных изменен

### Исправление ошибок в данных

**Столбец `children`.**

Проверим сколько отрицательных значений имеется.

In [22]:
clients[clients['children'] < 0]['children'].unique()

array([-1], dtype=int64)

Другие значения отсутствуют. Скорее всего это техническая ошибка, которую следует исправить. Берем значение по модулю.

In [23]:
clients['children'] = clients['children'].abs()

In [24]:
clients[clients['children'] < 0]['children'].unique()

array([], dtype=int64)

Теперь у нас нет отрицательного количества детей

Также можно заметить, что максимальное число детей - 20.
Сгруппируем семьи по количеству детей.

In [25]:
clients['children'].value_counts()

0     14149
1      4865
2      2055
3       330
20       76
4        41
5         9
Name: children, dtype: int64

20 детей встречайтся 76 раз. Это кажется странным, так как больше нет семей, где более 5 детей и менее 20.
Предполагаю, что это ошибка при сборе данных: изначально стоял ноль, но при вводе новых данных (2) не был удален.
Поменяем число детей на 2.

In [26]:
clients['children'] = clients['children'].replace(20, 2)

In [27]:
clients['children'].value_counts()

0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

**Столбец `dob_years`.**

Найдем количество клиентов, у которых в значении возраста проставлено 0.

In [28]:
clients.value_counts(clients['dob_years'] == 0)

dob_years
False    21424
True       101
dtype: int64

У 101 клиента неверно проставлен возраст. Предполагаю, что ошибка случайна. 


Найдем минимальное значение в колонке после 0.

In [29]:
clients[(clients['dob_years'] > 0) & (clients['dob_years'] < 20)].value_counts('dob_years')

dob_years
19    14
dtype: int64

Заменим все нулевые значения в колонке на 17 лет. Так как минимальное после 0 в наших данных это 19 лет, мы всего лишь сдвинем минимальный возраст на более разумную границу.

In [30]:
clients['dob_years'].replace(0, 17, inplace = True)

Проверим изменился ли минимальный возраст.

In [31]:
clients['dob_years'].min()

17

**Столбец `days_employed`.**

Максимально возможный стаж для человека возрастом 90 лет, если он работал с 18 лет:


In [32]:
f'{(90 - 18) * 365} дней.'

'26280 дней.'

Проверим есть ли в данном столбце значения большие 0, но при этом не превосходящие 26280 дней.

In [33]:
clients[(clients['days_employed'] >= 0) & (clients['days_employed'] < 26280)]

Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose]
Index: []

Таких значений нет, следовательно, в столбце указаны отрицательные знаения или значения, превосходящие человеческие возможности.

Можно предположить, что в данных произошла ошибка, при которой данные стали отрицательными или при сборе информации были введены неверные значения.

Приведем отрицательные значения к положительному путем отброса минуса и заменим экстремальные значения максимально возможным стажем.

Напишем функцию, которая высчитывает максимальный рабочий стаж.


In [34]:
def max_employed(dataframe):
    '''Функция выполняет поиск максимального трудового стажа.
    Если стаж больше максимального, то
    функция возвращает максимально возможный стаж,
    основываясь на возраст человека, в противном случае
    возвращает положительное значение исходного трудового стажа
    '''
    current_days_employed = dataframe['days_employed']
    if dataframe['dob_years'] > 65:
        max_years = 65
    else:
        max_years = dataframe['dob_years']
            
    
    max_days_employed = (max_years - 14) * 365

    if current_days_employed > max_days_employed:
        return max_days_employed
    elif current_days_employed < 0:
        return current_days_employed * -1
    else:
        return current_days_employed

In [35]:
clients['days_employed'] = clients.apply(max_employed, axis = 1)

Проверим изменения

In [36]:
clients.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   21525.000000  21525.000000  21525.000000   
mean       0.479721    4799.033728     43.373148      0.817236   
std        0.755528    5853.080194     12.351630      0.548138   
min        0.000000      24.000000     17.000000      0.000000   
25%        0.000000     974.000000     33.000000      1.000000   
50%        0.000000    2008.000000     42.000000      1.000000   
75%        1.000000    5293.000000     53.000000      1.000000   
max        5.000000   18615.000000     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  2.152500e+04  
mean           0.972544      0.080883  1.652248e+05  
std            1.420324      0.272661  9.804367e+04  
min            0.000000      0.000000  2.066700e+04  
25%            0.000000      0.000000  1.077980e+05  
50%            0.000000      0.000000  1.425940e+05  
75%            1.000000      0.000000  1.955490e+05  
max            4.000000      1.000000  2.265604e+06

### Обработка дубликатов


Посчитаем явные дубликаты в таблице

In [37]:
clients.duplicated().sum()

54

Удаление дубликатов

In [38]:
clients = clients.drop_duplicates().reset_index(drop=True)

Проверяем удалились ли все дубликаты

In [39]:
clients.duplicated().sum()

0

Теперь необходимо исправить неявные дубликаты в колонке `education`. 
Приведем данные к одному регистру.

In [40]:
clients['education'].unique()

array(['высшее', 'среднее', 'Среднее', 'СРЕДНЕЕ', 'ВЫСШЕЕ',
       'неоконченное высшее', 'начальное', 'Высшее',
       'НЕОКОНЧЕННОЕ ВЫСШЕЕ', 'Неоконченное высшее', 'НАЧАЛЬНОЕ',
       'Начальное', 'Ученая степень', 'УЧЕНАЯ СТЕПЕНЬ', 'ученая степень'],
      dtype=object)

In [41]:
clients['education'] = clients['education'].str.lower()

In [42]:
clients['education'].unique()

array(['высшее', 'среднее', 'неоконченное высшее', 'начальное',
       'ученая степень'], dtype=object)

Проверим `family_status` на наличие неявных дубликатов.

In [43]:
clients['family_status'].unique()

array(['женат / замужем', 'гражданский брак', 'вдовец / вдова',
       'в разводе', 'Не женат / не замужем'], dtype=object)

Тут все хорошо, но приведем все к нижнему регистру

In [44]:
clients['family_status'] = clients['family_status'].str.lower()

In [45]:
clients['family_status'].unique()

array(['женат / замужем', 'гражданский брак', 'вдовец / вдова',
       'в разводе', 'не женат / не замужем'], dtype=object)

### Вывод


- Были обработаны пропущенные значения по ежемесячному доходу на основе медиан для каждого типа занятости;
- Обработаны пропущенные значения в данных по трудовому стажу на основе медиан для каждой возрастной категории;
- Произведена замина типа данных на более удобный;
- Исправлены отрицательное и максимальное количество детей;
- Исправлены значения нулевого возраста;
- Максимальное значение трудового стажа приведено к реальному, основываясь на возрасте человека;
- Отрицательные значения трудового стажа приведены к положительному начению;
- Явные дубликаты были удалены, неявные приведены к общему регистру.


## Лемматизация

Столбец `purpose` содержит данные о цели кредита, описанные разными словами. Поэтому его следует лемматизировать и на основе его создать новый столбец с категориями.

Загрузим необходимую библиотеку. 
Выведем список уникальных целей.

In [46]:
from nltk.stem import SnowballStemmer 


In [47]:
clients['purpose'].value_counts()

свадьба                                   793
на проведение свадьбы                     773
сыграть свадьбу                           769
операции с недвижимостью                  675
покупка коммерческой недвижимости         662
покупка жилья для сдачи                   652
операции с жильем                         652
операции с коммерческой недвижимостью     650
покупка жилья                             646
жилье                                     646
покупка жилья для семьи                   638
строительство собственной недвижимости    635
недвижимость                              633
операции со своей недвижимостью           627
строительство жилой недвижимости          625
покупка недвижимости                      621
покупка своего жилья                      620
строительство недвижимости                619
ремонт жилью                              607
покупка жилой недвижимости                606
на покупку своего автомобиля              505
заняться высшим образованием      

Среди всех возможных целей получения кредита можно выделить следующие ключевые слова, которые отражают главную цель.

"жилье", "образование", "свадьба", "недвижимость", "автомобиль" ("жилье" отнесем к категории "недвижимость")

In [49]:
russian_stemmer = SnowballStemmer('russian')

In [50]:
def group_purpose(row):
    purpose_group = row['purpose']
    if 'недвиж'  in russian_stemmer.stem(purpose_group):
        return 'недвижимость'
    elif 'жил'  in russian_stemmer.stem(purpose_group):
        return 'недвижимость'
    elif 'авто' in russian_stemmer.stem(purpose_group):
        return 'автомобиль'
    elif 'образ' in russian_stemmer.stem(purpose_group):
        return 'образование'
    elif 'свад' in russian_stemmer.stem(purpose_group):
        return 'свадьба'
    else:
        return purpose_group

In [51]:
clients['purpose_grouped'] = clients.apply(group_purpose, axis = 1)


In [52]:
clients['purpose_grouped'].unique()

array(['недвижимость', 'автомобиль', 'образование', 'свадьба'],
      dtype=object)

In [54]:
clients.head(10)

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0          14235         53   среднее             1   
5         0            926         27    высшее             0   
6         0           2879         43    высшее             0   
7         0            152         50   среднее             1   
8         2           6929         35    высшее             0   
9         0           2188         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   
5  гражданский брак                 1      M   компаньон     0        255763   
6   женат / замужем                 0      F   компаньон     0        240525   
7   женат / замужем                 0      M   сотрудник     0        135823   
8  гражданский брак                 1      F   сотрудник     0         95856   
9   женат / замужем                 0      M   сотрудник     0        144425   

                      purpose purpose_grouped  
0               покупка жилья    недвижимость  
1     приобретение автомобиля      автомобиль  
2               покупка жилья    недвижимость  
3  дополнительное образование     образование  
4             сыграть свадьбу         свадьба  
5               покупка жилья    недвижимость  
6           операции с жильем    недвижимость  
7                 образование     образование  
8       на проведение свадьбы         свадьба  
9     покупка жилья для семьи    недвижимость

In [58]:
clients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21454 entries, 0 to 21453
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21454 non-null  int64 
 1   days_employed     21454 non-null  int64 
 2   dob_years         21454 non-null  int64 
 3   education         21454 non-null  object
 4   education_id      21454 non-null  int64 
 5   family_status     21454 non-null  object
 6   family_status_id  21454 non-null  int64 
 7   gender            21454 non-null  object
 8   income_type       21454 non-null  object
 9   debt              21454 non-null  int64 
 10  total_income      21454 non-null  int64 
 11  purpose           21454 non-null  object
 12  purpose_grouped   21454 non-null  object
dtypes: int64(7), object(6)
memory usage: 2.1+ MB


Проверим не вызвало ли это появление новых дубликатов.

In [55]:
clients.duplicated().sum()

17

Удалим дубликаты

In [56]:
clients = clients.drop_duplicates().reset_index(drop = True)

In [57]:
clients.duplicated().sum()

0

### Вывод

Мы видели, что в столбце purpose находились дублирующие причины, которые отличались формулировкой, поэтому причины стоило свести к одним и тем же категориям.


## Категоризация данных

### Катетеризация по количеству детей.

Разделим данные на группы "семьи без детей", "1-2 ребенка" и "многодетные". 
Напишем для этого функцию


In [68]:
def count_of_child(row):
    if row == 0:
        return 'семьи без детей'
    elif row == 1 or row == 2:
        return  '1-2 ребенка'
    else:
        return 'многодетные'

In [69]:
clients['type_of_family'] = clients['children'].apply(count_of_child)

### Категоризация по доходам

Для ответа на поставленные вопросы потребуется провести категоризацию по доходам.

Для этого воспользуемся квантилями 25%, 50% и 75%

1. 0% - 25%: низкая
2. 25% - 75%: средняя
3. 75% - 100%: высокая

In [61]:
clients['total_income'].describe().astype('int64')

count      21454
mean      165319
std        98187
min        20667
25%       107623
50%       142594
75%       195820
max      2265604
Name: total_income, dtype: int64

Напишем функцию для категоризации заработных плат



In [103]:
def classify_total_income(income):
    if income <= 107623:
        return 'низкая'
    elif 107623 < income <= 142594:
        return "средняя"
    else:
        return "высокая"

In [104]:
clients['total_income_status'] = clients['total_income'].apply(classify_total_income)

### Категоризация по возрасту

Также проведем категоризацию по возрасту.

In [74]:
clients['dob_years'].describe().astype('int64')

count    21454
mean        43
std         12
min         17
25%         33
50%         42
75%         53
max         75
Name: dob_years, dtype: int64

In [75]:
def classify_by_age(age):
    if age <= 30:
        return 'до 30 лет'
    elif 30 < age <= 45:
        return "30-45 лет"
    elif 45 < age <= 65:
        return "45-65 лет"
    else:
        return "старше 65 лет"

In [76]:
clients['dob_years_cat'] = clients['dob_years'].apply(classify_by_age)

In [77]:
clients.head()

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0          14235         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose purpose_grouped total_income_status  \
0               покупка жилья    недвижимость             высокая   
1     приобретение автомобиля      автомобиль             средняя   
2               покупка жилья    недвижимость             высокая   
3  дополнительное образование     образование             высокая   
4             сыграть свадьбу         свадьба             высокая   

    type_of_family dob_years_cat  
0      1-2 ребенка     30-45 лет  
1      1-2 ребенка     30-45 лет  
2  семьи без детей     30-45 лет  
3      многодетные     30-45 лет  
4  семьи без детей     45-65 лет

### Вывод

Теперь, основываясь на полученных данных, мы сможем изучить влияет ли какая-то категория на вероятность задолженности по кредиту



## Проверка гипотез

### Есть ли зависимость между наличием детей и возвратом кредита в срок?


In [93]:
depending_on_children = clients.groupby(['type_of_family'])['debt'].agg(['count', 'sum'])
depending_on_children['%'] = round(depending_on_children['sum'] / depending_on_children['count'] * 100, 2)
depending_on_children.sort_values(['%']).reset_index()

type_of_family  count   sum     %
0  семьи без детей  14091  1063  7.54
1      многодетные    380    31  8.16
2      1-2 ребенка   6983   647  9.27

Вероятность клиента принадлежащего к группе "многодетные" стать должником составляет 8.16%, а у семей, где  "1-2 ребенка" - 
9.27%. 

Вероятность клиента из группы "семьи без детей" стать должником составляет 7.54%

### Вывод

С увеличением количества детей мы видим увеличение количества просроченных задолженностей, хотя люди с 3 детьми чаще платят в срок чем люди с 1-2 детьми. 

Бездетные, как правило реже просрачивают оплату по кредиту, чем люди с детьми.

### Есть ли зависимость между семейным положением и возвратом кредита в срок?


In [115]:
depending_on_family_status = clients.groupby(['family_status'])['debt'].agg(['count', 'sum'])
depending_on_family_status['%'] = round(depending_on_family_status['sum'] / depending_on_family_status['count'] * 100, 2)
depending_on_family_status.sort_values(['%']).reset_index()

family_status  count  sum     %
0         вдовец / вдова    959   63  6.57
1              в разводе   1195   85  7.11
2        женат / замужем  12339  931  7.55
3       гражданский брак   4151  388  9.35
4  не женат / не замужем   2810  274  9.75

### Вывод

Не состоящие и не бывавшие в браке наиболее склонны к задержке выплат, в то же время вдовцы/вдовы и разведеные меньше всех склонны к задолженностям по кредиту.

### Есть ли зависимость между уровнем дохода и возвратом кредита в срок?


In [106]:
depending_on_income = clients.groupby(['total_income_status'])['debt'].agg(['count', 'sum'])
depending_on_income['%'] = round(depending_on_income['sum'] / depending_on_income['count'] * 100, 2)
depending_on_income.sort_values(['%']).reset_index()

total_income_status  count  sum     %
0             высокая  10611  831  7.83
1              низкая   5364  427  7.96
2             средняя   5479  483  8.82

### Вывод

Люди с высоким доходом меньше с меньшей вероятностью окажутся в должниках, как и люди с низким доходом.
В то время как наибольшая вероятность 8.82% стать должником у людей, получающих среднюю зарплату.

Это можно объяснить тем, что люди с низким доходам лучше распоряжаются своими средствами и умеют контролировать свой бюджет.

### Как разные цели кредита влияют на его возврат в срок?


In [108]:
depending_on_purpose = clients.groupby(['purpose_grouped'])['debt'].agg(['count', 'sum'])
depending_on_purpose['%'] = round(depending_on_purpose['sum'] / depending_on_purpose['count'] * 100, 2)
depending_on_purpose.sort_values(['%']).reset_index()

purpose_grouped  count  sum     %
0    недвижимость  10811  782  7.23
1         свадьба   2324  186  8.00
2     образование   4013  370  9.22
3      автомобиль   4306  403  9.36

### Вывод

Чаще всего платят в срок с целью кредита недвижимость.
Наибольшая вероятность 9.36% стать должником есть у людей, получающих кредитные средства на операции с автомобилем, и 9.22% на образование.



## Общий вывод

В ходе исследования надёжности заемщиков были изучены полученные данные, обработаны и заполнены пропуски, удалены дубликаты и клиенты были объединены в группы по разным параметрам.

По итогам проделанной работы можно сделать вывод, что семеное положение и количество детей влияет на факт погашения кредита в срок.

Идеальный заемщик кредита это: 
 - состоящий или состоящий ранее в официально зарегистрированном браке, 
 - без детей, 
 - зарабатывает больше 107 тыс. рублей в месяц, 
 - приобретатель недвижимости.

1. Наличие и количество детей повышает риск невыплат по кредиту
2. Люди официально состоящие или состоявшие ранее в браке более ответственны
3. Люди со средним доходом являются более безопасными клиентами, но необходимо тщательно проверять достоверность представленных документов.
4. Кредиты на недвижимость являются наиболее взвешенными, а покупка автомобилей наиболее импульсивными покупками.